### DISCUSSION TITLE
PageObject member function compress_content_streams

### DISCUSSION THREAD

**Discussion by j-t-1** [py-pdf/pypdf#2485] [Q&A] [answered]
```python3
from pypdf import PdfReader, PdfWriter

reader = PdfReader("input.pdf", "rb")
writer = PdfWriter()

for page in reader.pages:
    writer.add_page(page)

for page in writer.pages:
    page.compress_content_streams()

with open("output.pdf", "wb") as fd:
    writer.write(fd)
```

1. This works, but the links in the contents page of the PDF do not; is there an obvious reason for this?
2. For the function compress_content_streams, are there streams that are non-content that are specifically excluded?
3. Is there a better or more Pythonic way of doing the above, for example a zip or something else to make it a single for loop?

I tried this but the content page links also do not work:

```python3
from pypdf import PdfReader, PdfWriter

reader = PdfReader("big-old-file.pdf")
writer = PdfWriter()

for page in reader.pages:
    writer.add_page(page)

writer.add_metadata(reader.metadata)

with open("smaller-new-file.pdf", "wb") as fp:
    writer.write(fp)
```

https://github.com/py-pdf/pypdf/blob/main/docs/user/file-size.md

**Comment by stefan6419846** [comment, marked as answer]
You probably want to use `clone_from`:

```python3
writer = PdfWriter(clone_from="input.pdf")

for page in writer.pages:
    page.compress_content_streams()

writer.write("output.pdf")
```

While there might be more Pythonic constructs to simplify the loop (for example using list comprehension or `map`), I think that this just makes the action less obvious.

**Reply by j-t-1** [reply]
Thanks, this works (only had to change the last line to writer). I was unaware of the existence of the clone_from argument.

I share the following concern with @lababidi in the interesting 1604 discussion:

> One of the concerns I have is that writing a PDF after it has been read results in many changes in the PDF especially when nothing has been changed.

The clone_from argument seems to be a way of somewhat bypassing this problem.
